File: framework/languages/java/e2e-kotlin/RegistryMessaging/Client/src/main/kotlin/systems/zlink/e2e/kotlin/registrymessaging/client/Scenarios/RmC1RequestSendScenario.kt
package systems.zlink.e2e.kotlin.registrymessaging.client.Scenarios

import java.util.UUID
import systems.zlink.e2e.kotlin.registrymessaging.client.Support.HttpJson
import systems.zlink.e2e.kotlin.registrymessaging.client.Support.ScenarioAssert
import systems.zlink.e2e.kotlin.registrymessaging.shared.EvidenceWaitRequest
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileCommand
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileReply
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileRequest

object RmC1RequestSendScenario {
    fun run(providerA: HttpJson, providerB: HttpJson) {
        val reply = providerA.post<ProfileReply>("/profile/request", ProfileRequest("rm-c1-request"))
        ScenarioAssert.that(reply.value == "profile:rm-c1-request", "RM-C1 request reply mismatch.")

        val commandId = "cmd-${UUID.randomUUID().toString().replace("-", "")}"
        providerA.post<Map<String, Any>>("/profile/command", ProfileCommand(commandId))
        val evidence = providerA.post<List<String>>("/evidence/wait", EvidenceWaitRequest(commandId)) +
            providerB.post<List<String>>("/evidence/wait", EvidenceWaitRequest(commandId))
        ScenarioAssert.that(evidence.any { it.contains("profile-request|") && it.contains("rm-c1-request") }, "RM-C1 request evidence missing.")
        ScenarioAssert.that(evidence.any { it.contains("profile-command|") }, "RM-C1 send evidence missing.")
        println("scenario RM-C1 passed")
    }
}

File: framework/languages/dotnet/e2e/LocationMessaging/Client/Scenarios/RmC1RequestSendScenario.cs
// Verifies RM-C1 Request Send behavior.
using LocationMessaging.Client.Support;
using LocationMessaging.Shared;
using Zlink.HttpClient;

namespace LocationMessaging.Client.Scenarios;

// RM-C1 verifies the normal request/reply and one-way send paths for a
// location-store auto-connected profile channel.
internal static class RmC1RequestSendScenario
{
    public static async Task RunAsync(ZLinkHttpClient providerA, ZLinkHttpClient providerB)
    {
        var reply = (await providerA.Post("/profile/request")
            .Body(new ProfileReq("rm-c1-request"))
            .Async<ProfileRes>()).Body;
        ZlinkStreamAssert.Ensure(reply.Value == "profile:rm-c1-request", "RM-C1 request reply mismatch.");

        var commandId = $"cmd-{Guid.NewGuid():N}";
        await providerA.Post("/profile/command")
            .Body(new ProfileMsg(commandId))
            .Async<object>();

        var evidence = await WaitForProviderEvidenceAsync(providerA, providerB, commandId);
        ZlinkStreamAssert.Ensure(
            evidence.Any(line => line.Contains("profile-request|", StringComparison.Ordinal)
                                 && line.Contains("rm-c1-request", StringComparison.Ordinal)),
            "RM-C1 request evidence missing.");
        ZlinkStreamAssert.Ensure(
            evidence.Any(line => line.Contains("profile-command|", StringComparison.Ordinal)),
            "RM-C1 send evidence missing.");
    }

    private static async Task<string[]> WaitForProviderEvidenceAsync(
        ZLinkHttpClient providerA,
        ZLinkHttpClient providerB,
        string commandId)
    {
        // ChannelName send is select-one (spec 11 §2): the command lands on
        // exactly one member, so the wait completes on whichever provider
        // received it and the other side contributes its current snapshot.
        var wait = new EvidenceWaitReq($"command={commandId}");
        var providerAEvidence = providerA.Post("/evidence/wait").Body(wait).Async<string[]>().AsTask();
        var providerBEvidence = providerB.Post("/evidence/wait").Body(wait).Async<string[]>().AsTask();
        var winner = await Task.WhenAny(providerAEvidence, providerBEvidence);
        var other = ReferenceEquals(winner, providerAEvidence) ? providerB : providerA;
        var otherSnapshot = (await other.Get("/evidence").Async<string[]>()).Body;
        return (await winner).Body.Concat(otherSnapshot).ToArray();
    }
}
